js/ui/components/clubhouse.js

```diff
@@ -0,0 +1,323 @@
+    Extends: MessageTray.NotificationBanner,
+
+    _init: function(notification) {
+        this.parent(notification);
+
+        // We don't have an "unexpanded" state for now
+        this.expanded = true;
+        this._actionBin.visible = true;
+
+        this.actor.can_focus = true;
+        this.actor.track_hover = true;
+
+        // Override the style name because this is a not a regular notification
+        this.actor.remove_style_class_name('notification-banner');
+        this.actor.add_style_class_name('clubhouse-quest-view');
+        this._iconBin.add_style_class_name('clubhouse-quest-view-icon-bin');
+
+        // Always wrap the body's text
+        this.bodyLabel.actor.clutter_text.line_wrap = true;
+        this.bodyLabel.actor.add_style_class_name('clubhouse-quest-view-label');
+
+        this._closeButton.connect('clicked', Lang.bind(this, function() {
+            this.emit('dismiss');
+        }));
+    },
+});
+Signals.addSignalMethods(ClubhouseNotificationBanner.prototype);
+
+var ClubhouseNotification = new Lang.Class({
+    Name: 'ClubhouseNotification',
+    Extends: NotificationDaemon.GtkNotificationDaemonNotification,
+
+    _init: function(source, notification) {
+        this._currentIconPath = null;
+
+        let gicon = notification.icon;
+        if (gicon) {
+            let newGicon = this._getClubhouseIcon(Gio.icon_deserialize(gicon));
+            notification.icon = newGicon.serialize();
+        }
+
+        // We don't set up a title for now
+        notification.title = GLib.Variant.new('s', '');
+
+        this.parent(source, notification);
+    },
+
+    _getClubhouseIcon: function(gicon) {
+        let iconPath = gicon.to_string();
+
+        if (iconPath.startsWith('/app/'))
+            iconPath = iconPath.substring(5);
+
+        // Check if this icon is already loaded, in which case we return null so it won't
+        // be updated.
+        if (this._currentIconPath == iconPath)
+            return null;
+
+        this._currentIconPath = iconPath;
+        let iconFile = this._getClubhouseAsset(iconPath);
+
+        return Gio.FileIcon.new(iconFile);
+    },
+
+    _getClubhouseAsset: function(assetSubpath) {
+        let installationPath = Flatpak.Installation.new_user(null).get_path().get_path();
+        let assetPath = GLib.build_filenamev([installationPath, 'app', CLUBHOUSE_ID, 'current',
+                                               'active', 'files', assetSubpath]);
+        let assetFile = Gio.File.new_for_path(assetPath);
+
+        if (assetFile.query_exists(null))
+            return assetFile;
+
+        return null;
+    },
+
+    createBanner: function() {
+        return new ClubhouseNotificationBanner(this);
+    },
+
+    updateFromParams: function(params) {
+        // Update actions/buttons
+        this.actions = [];
+
+        if (this._buttonBox) {
+            this._buttonBox.destroy();
+            this._buttonBox = null;
+        }
```

Comment:
It looks like we could call `this.update()` with `clear=true` to get this?

Reply:
The problem with that is that the icon will also be cleared, which means that it'd always have to load the icon again. We can try to work around this by keeping the icons cached (if we care at all).

Reply:
I have changed this too.